contracts/SDUtilityPool.sol

```diff
@@ -0,0 +1,771 @@
+            revert SDTransferFailed();
+        }
+        uint256 cTokenToMint = (sdAmount * DECIMAL) / exchangeRate;
+        delegatorCTokenBalance[msg.sender] += cTokenToMint;
+        cTokenTotalSupply += cTokenToMint;
+
+        emit Delegated(msg.sender, sdAmount, cTokenToMint);
+    }
+
+    function _requestWithdraw(uint256 _sdAmountToWithdraw, uint256 cTokenToBurn) internal returns (uint256) {
+        if (requestIdsByDelegatorAddress[msg.sender].length + 1 > maxNonRedeemedDelegatorRequestCount) {
+            revert MaxLimitOnWithdrawRequestCountReached();
+        }
+        sdRequestedForWithdraw += _sdAmountToWithdraw;
+        delegatorWithdrawRequests[nextRequestId] = DelegatorWithdrawInfo(
+            msg.sender,
+            cTokenToBurn,
+            _sdAmountToWithdraw,
+            0,
+            block.number
+        );
+        requestIdsByDelegatorAddress[msg.sender].push(nextRequestId);
+        emit WithdrawRequestReceived(msg.sender, nextRequestId, _sdAmountToWithdraw);
+        nextRequestId++;
+        return nextRequestId - 1;
+    }
+
+    function _utilize(address utilizer, uint256 utilizeAmount) internal {
+        /* Verify `accrualBlockNumber` block number equals current block number */
+        if (accrualBlockNumber != block.number) {
+            revert AccrualBlockNumberNotLatest();
+        }
+        if (getPoolAvailableSDBalance() - sdRequestedForWithdraw < utilizeAmount) {
+            revert InsufficientPoolBalance();
+        }
+        uint256 accountUtilizePrev = _utilizerBalanceStoredInternal(utilizer);
+
+        utilizerData[utilizer].principal = accountUtilizePrev + utilizeAmount;
+        utilizerData[utilizer].utilizeIndex = utilizeIndex;
+        totalUtilizedSD = totalUtilizedSD + utilizeAmount;
+        ISDCollateral(staderConfig.getSDCollateral()).depositUtilizedSD(utilizer, utilizeAmount);
+        emit SDUtilized(utilizer, utilizeAmount);
+    }
+
+    function _repay(address utilizer, uint256 repayAmount) internal {
+        /* Verify `accrualBlockNumber` block number equals current block number */
+        if (accrualBlockNumber != block.number) {
+            revert AccrualBlockNumberNotLatest();
+        }
+
+        /* We fetch the amount the utilizer owes, with accumulated fee */
+        uint256 accountUtilizePrev = _utilizerBalanceStoredInternal(utilizer);
+
+        uint256 repayAmountFinal = (repayAmount == type(uint256).max || repayAmount > accountUtilizePrev)
+            ? accountUtilizePrev
+            : repayAmount;
+
+        if (!IERC20(staderConfig.getStaderToken()).transferFrom(msg.sender, address(this), repayAmountFinal)) {
+            revert SDTransferFailed();
+        }
+        if (!staderConfig.onlyStaderContract(msg.sender, staderConfig.SD_COLLATERAL())) {
+            uint256 feeAccrued = accountUtilizePrev -
+                ISDCollateral(staderConfig.getSDCollateral()).operatorUtilizedSDBalance(utilizer);
+            if (repayAmountFinal > feeAccrued) {
```

Comment:
Why should the utility pool care about how much was utilized vs interest-accrued for an operator? Feels like this logic should be part of SDCollateral contract

Reply:
if not SD_Collateral caller, you not only reduce utilizedPosition but also increase selfBondedPosition by that number which wont be the case with SD_Collateral caller